### ISSUE TITLE
unable to replace RouterDelegate or Router with Nav 2.0

### ISSUE THREAD

**Issue by dgaedcke** [flutter/flutter#69058]
Perhaps this is a documentation issue, but since both Router and RouterDelegate are FINAL, it seems there is no way to swap them out without reconstructing the whole MaterialApp.

And that seems very cumbersome and limited to me.  

For example, I have a package that handles all of the Social Login (Federated Auth) boilerplate and wrapped SDK's and all of that repetitive nonsense.  After it does it's work (auth is complete), I want it to clean up after itself and put the APP-LEVEL RouterDelegate in place.

I don't currently see a way to do this.   Please tell me what I'm missing.

**Comment by darshankawar** [comment]
@dgaedcke 
Please go through [this](https://github.com/flutter/flutter/issues/45938) issue and see if any of the discussion held around `RouterDelegate` and the Medium post shared in it helps you to answer your question.

**Comment by dgaedcke** [comment]
I went through that entire issue prior to posting this, but just to confirm I didn't miss anything, I reviewed it again.  And I've also been through that Medium post twice as well.

I feel like this I understand the problem fairly well ... my issue is not an educational one ... the paradox is pretty clear:

The docs clearly imply you can swap delegates:
`If the **identity of the Router widget's delegates change** while an asynchronous operation is in progress, to keep matters simple, all active asynchronous operations will have their results discarded. It is generally considered unusual for these delegates to change during the lifetime of the Router.`

And the implementation has **BOTH the Router and RouterDelegate marked as FINAL** ... this means they cannot be replaced without rebuilding the ENTIRE Material app ... and that seems crazy to me.

I'm fairly senior and I strongly suspect this is an oversight that someone on the team needs to review ASAP before Nav 2.0 is adopted in wide usage.

**Comment by PixelToast** [comment]
Hi @dgaedcke 

Re-building the MaterialApp is the intended way of switching out a routerDelegate (Widgets are immutable structures, so its fields must be final), this shouldn't be a very cumbersome task for a StatefulWidget.

**Comment by dgaedcke** [comment]
That's a bit surprising but I'll take you at your word.
I supposed I could do it just the same with a future-builder if I like?

In any case, please see the docs I copied above:
`If the identity of the Router widget's delegates change while an asynchronous operation is in progress, to keep matters simple`

That needs to be corrected because it implies something inconsistent with your position ...

**Comment by PixelToast** [comment]
Documentation often conflates `Widget`s (the immutable description), `Element`s (the instantiation), and `State`s (stateful delegates) so I understand why that statement might be confusing. From a `_RouterState`'s point of view the value of `widget.routerDelegate` can change between rebuilds, which is what the comment refers to.

**Comment by chunhtai** [comment]
The `router` in this paragraph
`If the **identity of the Router widget's delegates change** while an asynchronous operation is in progress, to keep matters simple, all active asynchronous operations will have their results discarded. It is generally considered unusual for these delegates to change during the lifetime of the Router.`

is referring to the element of Router widget. We usually say `lifetime of the widget` is actually the `lifetime of the element created from the widget`.

otherwise, as @PixelToast  said, the way to swap out the RouterDelegate is to rebuild `MaterialApp.router`.

**Comment by dgaedcke** [comment]
If the only way to replace the delegate is to rebuild the MaterialApp, then that implies that a NEW router has been created.  When would that not imply a new Router Element as well, since the whole tree (and related scope) has changed?

In any case, I appreciate the feedback!!  I'll leave it to you to clarify the docs if needed, or to me to study more if not.  Thanks for the help.  Closing this issue.

**Comment by chunhtai** [comment]
It will not create a new router element. If it determine the new router widget and the old router widget are compatible (Runtime type and widget key are the same), it will update the existing router element with the new widget through didUpdateWidget. So in this case, it will not create a new router element.

**Comment by dgaedcke** [comment]
So I need to verify that either BOTH material apps use the SAME key (is this for the navigator??) or that NEITHER use any key ... is that correct??

Thanks for the clarification!!

**Comment by chunhtai** [comment]
The router created in MaterialApp does not have key, so you just need to make sure the material app you create either use same key or no key. This will make sure the material app subtree will not get destroyed as a result of material app getting dispose and recreate.

**Comment by dgaedcke** [comment]
I thought about this some more, and it seems I DO want the material app subtree to be destroyed.  

I'll be switching from an Authentication flow, that uses it's own Router/Nav/Delegate for a unique stack of (auth-related) pages, and then once that's done, the next MaterialApp will have it's own home-screen and unique, totally new destination pages.

In this case, I don't see any subtree worth preserving so I'm guessing that the key issue is fairly moot in this use-case.

**Comment by chunhtai** [comment]
@dgaedcke I don't see why you want to destroy material app subtree in this case, you just need to swap out the page list in the Navigator you created in the RouterDelegate, or like you said, swapping out the RouterDelegate, but doing that alone will not destroy the material app.

**Comment by dgaedcke** [comment]
Interesting!!!  Thanks for the feedback ... this is making it clear that I don't really know what the "subtree" is because if it's not the rendered routes/pages/widgets on the screen, then it must be some internal element structures of the material app.   I'll do some more reading and I'll follow your advice and use keys for both MaterialApps.

I know it's off base for this thread (even if it were not closed), but I've been struggling for 3 days to find out why the code below is NEVER calling the "build" method of my router delegate ... so I'm never getting a Navigator .... just a blank white screen.  I've got a print statement in my "addListener" override, and that IS getting called (so router is constructed and attached), but the "build" method is not ... I'm a bit baffled if you have any ideas.  I tried to force-call "notifyListeners()" but it seems Flutter does not like such things while main is still running (ie before the even-loop is started).  All feedback most appreciated.

  @override
  Widget build(BuildContext context) {
    print('Running build of SLH Material App.. ${SvcLocator.routerDelegate}');
    return Provider<SlhFlowDispatcher>.value(
      value: SvcLocator.actionDispatcher,
      child: MaterialApp.router(
        key: appKey, // this may be null?
        title: appTitle,
        routerDelegate: SvcLocator.routerDelegate,
        routeInformationParser: SlhRouteParser(),
        // routeInformationProvider: PlatformRouteInformationProvider(
        //   initialRouteInformation: RouteInformation(location: '/'),
        // ),
      ),
    );
  }
